pkg/mailboxwatch/interface.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mailboxwatch
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	restclient "k8s.io/client-go/rest"
+	upstreamcache "k8s.io/client-go/tools/cache"
+
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	kcpinformers "github.com/kcp-dev/apimachinery/v2/third_party/informers"
+	kcpscopedclientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpapiinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	tenancyv1a1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+)
+
+// NewSharedInformer constructs a shared index informer on objects of a given kind in mailbox workspaces.
+// It follows the usual pattern for a constructor of informers except for an additional parameter
+// at the front, which it uses to keep appraised of the mailbox workspaces.
+// The ListerWatcher can be constructed using `upstreamcache.NewListWatchFromClient`.
+func NewSharedInformer(
+	mailboxWorkspacePreInformer tenancyv1a1informers.WorkspaceInformer,
+	listerWatcher upstreamcache.ListerWatcher,
+	exampleObject runtime.Object,
+	defaultEventHandlerResyncPeriod time.Duration,
+	indexers upstreamcache.Indexers,
+) kcpcache.ScopeableSharedIndexInformer {
+	if true {
+		panic("not implemented yet")
+	}
+	// Implementation outline:
+	// Use the informer on mailbox workspaces to stay appraised of the logicalcluster.Name for
+	// each mailbox workspace qua logical cluster.
```

Comment:
I suppose this is a typo ('`workspace qua logical`) but I cannot infer what this was intended to mean.

Reply:
Below is an example display of a Workspace object.  You can see two logical clusters mentioned in there.  The containing one, in the annotations, and the one that the Workspace represents, in the spec. 

```console
(base) mspreitz@mjs12 kcp % kubectl get Workspace edge -o yaml
apiVersion: tenancy.kcp.io/v1alpha1
kind: Workspace
metadata:
  annotations:
    experimental.tenancy.kcp.io/owner: '{"username":"kcp-admin"}'
    internal.tenancy.kcp.io/cluster: dto88oazto05t06o
    internal.tenancy.kcp.io/shard: 1pfxsevk
    kcp.io/cluster: root
  creationTimestamp: "2023-03-20T18:49:25Z"
  finalizers:
  - core.kcp.io/logicalcluster
  generation: 2
  labels:
    tenancy.kcp.io/phase: Ready
  name: edge
  resourceVersion: "836"
  uid: 1c72205c-1bb6-4f54-9d82-10e941cbde8a
spec:
  URL: https://${myip}:6443/clusters/dto88oazto05t06o
  cluster: dto88oazto05t06o
  type:
    name: organization
    path: root
status:
  conditions:
  - lastTransitionTime: "2023-03-20T18:49:26Z"
    status: "True"
    type: WorkspaceInitialized
  - lastTransitionTime: "2023-03-20T18:49:25Z"
    status: "True"
    type: WorkspaceScheduled
  phase: Ready
```